### PR TITLE
Spill registers on the main stack on compartment call.

### PR DESCRIPTION
**Pull request by davidchisnall**
Originally, we spilled the callee-save registers onto the trusted stack. This had some nice properties but means that we need 16 bytes per trusted stack frame that are not usable for anything else when higher up the call stack.

Moving the spills to the main stack introduces three potential attacks:

 - If there is not sufficient space on the stack, the switcher will trap on entry.
 - If csp does not have store-capability permission, the switcher will trap on entry.
 - If csp does not have load-capability permission, the switcher will trap when popping a trusted stack frame.

These are all addressed by the logic that makes the switcher do a forced unwind if it detects a trap within the switcher's PCC.  We already have tests for all missing permissions in CSP.  This was almost certainly not safe to do before we added that defence in the switcher, but is now.

This makes no difference to the deepest compartment call on a thread, but if the memory is allocated to the stack instead of the trusted stack then each higher frame can have 16 bytes more stack space per unused trusted stack frame than previously.  This also makes it possible to support RV32I (which has a lot more callee-save registers) with some simple changes.

### CODE REVIEW

**Comment by davidchisnall** [comment]
We have tests for many variations on malicious stacks from the caller and callee.

**Comment by davidchisnall** [comment]
Looking at the code again, there's no reason that pcc and cgp should remain on the trusted stack.  We can probably move them to the normal stack as well.

File: sdk/core/switcher/trusted-stack-assembly.h [hunk 1]
```diff
@@ -39,19 +39,17 @@ EXPORT_ASSEMBLY_OFFSET(TrustedStack, mshwmb, (17 * 8) + 4)
 // unwinding information, and then a single trusted stack frame used for the
 // unwind state of the initial thread. (7 * 8) is the size of TrustedStackFrame
```
Comment (nwf-msr) [inline sdk/core/switcher/trusted-stack-assembly.h:40]: Comment wants updating, too?

File: compile_flags.txt [hunk 2]
```diff
@@ -31,7 +31,7 @@ riscv32-unknown-unknown
 -DDEVICE_EXISTS_shadow
 -DDEVICE_EXISTS_uart
 -DDEVICE_EXISTS_clint
--DCHERIOT_LOADER_TRUSTED_STACK_SIZE=208
+-DCHERIOT_LOADER_TRUSTED_STACK_SIZE=192
```
[the remark this answers is not in the document]
Comment (davidchisnall) [reply compile_flags.txt:34]: You cannot add comments in this file.

File: sdk/xmake.lua [hunk 3]
```diff
@@ -639,7 +639,7 @@ target("cheriot.loader")
 		target:set('cherimcu.debug-name', "loader")
 		local config = {
 			-- Size in bytes of the trusted stack.
-			loader_trusted_stack_size = 224,
+			loader_trusted_stack_size = 208,
```
[the remark this answers is not in the document]
Comment (davidchisnall) [reply sdk/xmake.lua:642]: It's calculated by getting it wrong and then setting it to the value in the compiler error message.

File: sdk/core/switcher/trusted-stack-assembly.h [hunk 4]
```diff
@@ -39,19 +39,17 @@ EXPORT_ASSEMBLY_OFFSET(TrustedStack, mshwmb, (17 * 8) + 4)
...
 EXPORT_ASSEMBLY_OFFSET(TrustedStackFrame, cgp, 8)
 EXPORT_ASSEMBLY_OFFSET(TrustedStackFrame, csp, 16)
-EXPORT_ASSEMBLY_OFFSET(TrustedStackFrame, cs0, 24)
-EXPORT_ASSEMBLY_OFFSET(TrustedStackFrame, cs1, 32)
-EXPORT_ASSEMBLY_OFFSET(TrustedStackFrame, calleeExportTable, 40)
-EXPORT_ASSEMBLY_OFFSET(TrustedStackFrame, errorHandlerCount, 48)
-EXPORT_ASSEMBLY_SIZE(TrustedStackFrame, (8 * 8))
+EXPORT_ASSEMBLY_OFFSET(TrustedStackFrame, calleeExportTable, 24)
+EXPORT_ASSEMBLY_OFFSET(TrustedStackFrame, errorHandlerCount, 32)
+EXPORT_ASSEMBLY_SIZE(TrustedStackFrame, (8 * 6))
```
[the remark this answers is not in the document]
Comment (davidchisnall) [reply sdk/core/switcher/trusted-stack-assembly.h:52]: No, this value must be something that evaluates correctly in assembly.  If we could use `sizeof` here, we wouldn't need the size at all.

**Pull request merged**